libsnark/zk_proof_systems/plonk/srs.hpp

```diff
@@ -0,0 +1,171 @@
+- class for key pair (proving key & verification key) (TODO: not implemented)
+
+References:
+
+\[GWC19]:
+"Plonk: Permutations over lagrange-bases for oecumenical noninteractive
+arguments of knowledge", Ariel Gabizon, Zachary J. Williamson, and Oana
+Ciobotaru, Cryptology ePrint Archive, Report 2019/953, 2019,
+<https://eprint.iacr.org/2019/953>
+
+*****************************************************************************
+* @author     This file is part of libsnark, developed by SCIPR Lab
+*             and contributors (see AUTHORS).
+* @copyright  MIT license (see LICENSE file)
+*****************************************************************************/
+
+#ifndef PLONK_PPZKSNARK_SRS_HPP_
+#define PLONK_PPZKSNARK_SRS_HPP_
+
+namespace libsnark
+{
+/********************************** SRS ***********************************/
+
+// description of the circuit. contains only the number of gates for
+// now.
+template<typename ppT> struct circuit_t {
+    // number of gates in the analyzed circuit, denoted by "n" in
+    // [GWC19]
+    size_t num_gates;
+};
+
+//
+// A note on the distinction between an srs and a universal srs.
+//
+// A universal srs (usrs) is composed *only* of monomials i.e. encoded
+// powers of the secret value in the group generator. Therefore a usrs
+// is independent of any particular circuit.
+//
+// The (plain) srs is a specialization of the usrs for one particular
+// circuit and is derived from the usrs e.g. as
+//
+// usrs = <encoded powers of secret>
+// srs = (proving_key, verificataion_key) = derive(usrs, circuit_description)
+//
+
+// Universal srs (usrs). Contains secret encoded monomials with
+// maximum degree MAX_DEGREE and is so independent of any particular
+// circuit.
+template<typename ppT> class usrs
+{
+public:
+    /// Array of powers of secret \alpha, encoded in G1:
+    /// [1]_1, [\alpha]_1, [\alpha^2]_1, ..., [\alpha^{n+2}]_1
+    std::vector<libff::G1<ppT>> secret_powers_g1;
+
+    /// Array of powers of secret \alpha, encoded in G2:
+    /// [1]_2, [\alpha]_2
+    std::vector<libff::G2<ppT>> secret_powers_g2;
+
+    usrs(){};
+
+    usrs(
+        std::vector<libff::G1<ppT>> &&secret_powers_g1,
+        std::vector<libff::G2<ppT>> &&secret_powers_g2)
+        : secret_powers_g1(secret_powers_g1), secret_powers_g2(secret_powers_g2)
+    {
+    }
+
+    // for debug only
+    void derive_from_secret(const libff::Fr<ppT> secret);
+};
+
+// Plain (i.e. non-universal srs). Contains secret encoded monomials
+// with maximum degree equal to the number of gates of the analyzed
+// circuit + 2. Dependent on the circuit.
+template<typename ppT> class srs
+{
+public:
+    // description of the circuit. contains only the number of gates for
+    // now. the highest degree of the encoded power monomials will
+    // be num_gates+2.
+    circuit_t<ppT> circuit;
+    /// Array of powers of secret \alpha, encoded in G1:
+    /// [1]_1, [\alpha]_1, [\alpha^2]_1, ..., [\alpha^{n+2}]_1
```

Comment:
Ok here you seem to use the `///` comment syntax but this is not done consistently.

Reply:
See: https://www.doxygen.nl/manual/docblocks.html (we use `///` for comment blocks)